openedx/features/learner_analytics/views.py

```diff
@@ -119,3 +132,105 @@ def get_schedule(self, request, course_key):
+            weekly_course_activities = self.analytics_client.courses(course_key).activity()
+
+            if not weekly_course_activities or 'any' not in weekly_course_activities[0]:
+                return 0
+
+            # weekly course activities should only have one item
+            activities = weekly_course_activities[0]
+            cache.set(cache_key, activities, LearnerAnalyticsView.seconds_to_cache_expiration())
+
+        return activities['any']
+
+    def consecutive_weeks_of_course_activity_for_user(self, username, course_key):
+        """
+        Get the most recent count of consecutive days that a user has performed a course activity
+
+        Args:
+            username: Username
+            course_key: CourseKey
+        """
+        cache_key = 'learner_analytics_{username}_{course_key}_engagement_timeline'\
+            .format(username=username, course_key=course_key)
+        timeline = cache.get(cache_key)
+
+        if not timeline:
+            log.info('Engagement timeline for course {course_key} was not cached - fetching from Analytics API'
+                     .format(course_key=course_key))
+
+            # TODO: @jaebradley replace this once the Analytics client has an engagement timeline method
+            url = '{base_url}/engagement_timelines/{username}?course_id={course_key}'\
+                .format(base_url=settings.ANALYTICS_API_URL,
+                        username=username,
+                        course_key=urllib.quote_plus(course_key))
+            headers = {'Authorization': 'Token {token}'.format(token=settings.ANALYTICS_API_KEY)}
+            response = requests.get(url=url, headers=headers)
+            data = response.json()
+
+            if not data or 'days' not in data or not data['days']:
+                return 0
+
+            # Analytics API returns data in ascending (by date) order - we want to count starting from most recent day
+            data_ordered_by_date_descending = list(reversed(data['days']))
+
+            cache.set(cache_key, data_ordered_by_date_descending, LearnerAnalyticsView.seconds_to_cache_expiration())
+            timeline = data_ordered_by_date_descending
+
+        return LearnerAnalyticsView.calculate_week_streak(timeline)
+
+    @staticmethod
+    def calculate_week_streak(daily_activities):
+        """
+        Check number of weeks in a row that a user has performed some activity.
+
+        Regardless of when a week starts, a sufficient condition for checking if a specific week had any user activity
+        (given a list of daily activities ordered by date) is to iterate through the list of days 7 days at a time and
+        check to see if any of those days had any activity.
+
+        Args:
+            daily_activities: list of dictionaries containing activities and their counts
+        """
+        week_streak = 0
+        seven_day_buckets = [daily_activities[i:i + 7] for i in range(0, len(daily_activities), 7)]
```

Comment:
Have you tested this? It seems like it would run into problems once `i` gets close to `len(daily_activities)`.

Reply:
I checked it by doing

![image](https://user-images.githubusercontent.com/8136030/34223847-c4ba878c-e58e-11e7-8634-693b3c85ad25.png)

Let me know if I'm missing a test case